File: ffgraph/src/iohandler.c
#include "ffgraph.h"

#define BUFFER_SIZE (256 * 1024)                                // 256KiB
#define BUFFER_SIZE_BIG_FILES (1024 * 1024 * 4)                 // 4MiB
#define FILEHANDLE_MAX_BUFFERED_FILE_LENGTH (128 * 1024 * 1024) // 128MiB


static int iohandler_read(void* opaque, uint8_t* buf, int buf_size) {
    FileHandle_t* filehandle = opaque;

    int readed = filehandle_read(filehandle, buf, buf_size);

    // check if the end-of-file was reached
    if (readed < 0) return AVERROR_EOF;

    return readed;
}

static int64_t iohandler_seek(void* opaque, int64_t offset, int whence) {
    FileHandle_t* filehandle = opaque;
    if (whence == AVSEEK_SIZE) return filehandle_length(filehandle);

    whence &= ~AVSEEK_FORCE;

    int64_t new_offset = filehandle_seek(filehandle, offset, whence);

    return new_offset < 0 ? AVERROR_UNKNOWN : new_offset;
}


AVIOContext* iohandler_init(const FileHandle_t* filehandle) {
    FileHandle_t* filehnd = (FileHandle_t*)filehandle;

    // just in case
    filehandle_seek(filehnd, 0, SEEK_SET);

    int buffer_size;
    if (filehandle_length(filehnd) > FILEHANDLE_MAX_BUFFERED_FILE_LENGTH)
        buffer_size = BUFFER_SIZE_BIG_FILES;
    else
        buffer_size = BUFFER_SIZE;

    uint8_t* buffer = av_malloc(buffer_size);
    if (!buffer) {
        printf("iohandler_init() Can not allocate the buffer.\n");
        return NULL;
    }

    AVIOContext* avio_ctx = avio_alloc_context(
        buffer, buffer_size, 0, filehnd, iohandler_read, NULL, iohandler_seek
    );
    if (!avio_ctx) {
        printf("iohandler_init() call to avio_alloc_context() failed.\n");
        return NULL;
    }

    return avio_ctx;
}

void iohandler_destroy(AVIOContext** iohandle) {
    AVIOContext* avio_ctx = *iohandle;

    if (avio_ctx->buffer) av_free(avio_ctx->buffer);
    avio_context_free(iohandle);
}

File: src-desktop/ffgraph/iohandler.cs
using System;
using System.IO;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using Engine.Externals.FFGraphInterop.Interop;
using Engine.Externals.SoundBridge;

namespace Engine.Externals.FFGraphInterop;

internal unsafe class IOHandler : IDisposable {

    private const int BUFFER_SIZE = 256 * 1024;// 256KiB
    private const int BUFFER_SIZE_BIG_FILES = 1024 * 1024 * 4;// 4MiB
    private const int SOURCEHANDLE_MAX_BUFFERED_FILE_LENGTH = 128 * 1024 * 1024;// 128MiB

    private GCHandle src_hnd;
    private AVIOContext* avio_ctx;

    private IOHandler() { }


    private static ISourceHandle Recover(nint ptr) {
        return (ISourceHandle)GCHandle.FromIntPtr(ptr).Target;
    }


    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
    static int Read(nint opaque, byte* buf, int buf_size) {
        ISourceHandle sourcehandle = Recover(opaque);

        int readed = sourcehandle.Read(buf, buf_size);

        if (readed == 0)
            return FFmpeg.AVERROR_EOF;
        else if (readed < 0)
            return FFmpeg.AVERROR_UNKNOWN;
        else
            return readed;
    }

    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
    static long Seek(nint opaque, long offset, int whence) {
        ISourceHandle sourcehandle = Recover(opaque);
        if (whence == FFmpeg.AVSEEK_SIZE) return sourcehandle.Length();

        whence &= ~FFmpeg.AVSEEK_FORCE;

        int ret = sourcehandle.Seek(offset, (SeekOrigin)whence);

        if (ret > 0)
            return FFmpeg.AVERROR_UNKNOWN;
        else
            return sourcehandle.Tell();
    }


    public static IOHandler Init(ISourceHandle src_hnd) {
        // just in case
        src_hnd.Seek(0, SeekOrigin.Begin);

        int buffer_size;
        if (src_hnd.Length() > SOURCEHANDLE_MAX_BUFFERED_FILE_LENGTH)
            buffer_size = BUFFER_SIZE_BIG_FILES;
        else
            buffer_size = BUFFER_SIZE;

        byte* buffer;

        try {
            buffer = (byte*)FFmpeg.av_malloc((nuint)buffer_size);
        } catch {
            Logger.Error("IOHanlder::Init() Can not allocate the buffer.");
            return null;
        }

        IOHandler iohandler = new IOHandler();

        iohandler.src_hnd = GCHandle.Alloc(src_hnd, GCHandleType.Normal);
        nint ptr = GCHandle.ToIntPtr(iohandler.src_hnd);

        iohandler.avio_ctx = FFmpeg.avio_alloc_context(
            buffer, buffer_size, 0, ptr, &Read, null, &Seek
        );

        if (iohandler.avio_ctx == null) {
            Logger.Error("IOHanlder::Init() call to avio_alloc_context() failed.");
            iohandler.src_hnd.Free();
            return null;
        }

        return iohandler;
    }

    public void Dispose() {
        if (this.avio_ctx->buffer != null) FFmpeg.av_free(this.avio_ctx->buffer);

        fixed (AVIOContext** ptr = &avio_ctx) {
            FFmpeg.avio_context_free(ptr);
        }

        this.src_hnd.Free();
    }

    public AVIOContext* IOContext {
        get
        {
            return this.avio_ctx;
        }
    }

}

